proposed/logging-generator.md

```diff
@@ -0,0 +1,489 @@
+
+```csharp
+var receiverTemp = logger;
+var builder = CustomLoggerParamsBuilder.Create(
+    baseLength: 6,
+    formatHoleCount: 1,
+    receiverTemp, LogLevel.Information, out var builderIsValid);
+_ = builderIsValid &&
+    && builder.TryFormatBaseString("Hello ")
+    && builder.TryFormatInterpolationHole(name, "name");
+
+logger.Log<CustomLoggerParamsBuilder>(
+    LogLevel.Information,
+    new EventId(1),
+    builder, // the CustomLoggerParamsBuilder as log state
+    exception,
+    (s, e) => s
+);
+```
+
+
+
+On C# 10 and above. `CustomLoggerParamsBuilder` would be a custom [string interpolation builder](https://github.com/333fred/csharplang/blob/2550a43b391e844faaa0a2023f66489328a41612/proposals/improved-interpolated-strings.md) defined as:
+
+```csharp
+// The builder that will actually "build" the interpolated string"
+public struct CustomLoggerParamsBuilder : IReadOnlyList<KeyValuePair<string, object>>
+{
+    public static CustomLoggerParamsBuilder Create(int baseLength, int formatHoleCount, ILogger logger, LogLevel logLevel, out bool builderIsValid)
+    {
+        if (!logger.IsEnabled(logLevel))
+        {
+            builderIsValid = false;
+            return default;
+        }
+
+        builderIsValid = true;
+        return new CustomLoggerParamsBuilder(baseLength, formatHoleCount, logLevel);
+    }
+
+    private LogLevel _logLevelEnabled;
+
+    // returns the number of holes
+    public int Count => ... // depends on the type used to store key value pairs
+
+    public static readonly Func<CustomLoggerParamsBuilder, Exception?, string> FormatCallback = (builder, exception) => builder.ToString();
+
+    private CustomLoggerParamsBuilder(int baseLength, int formatHoleCount, LogLevel logLevelEnabled)
+    {
+        // Initialization logic
+        _logLevelEnabled = logLevelEnabled;
+    }
+
+    public bool TryFormatBaseString(string s)
+    {
+        // Store and format part as required
+        return true;
+    }
+
+    public bool TryFormatInterpolationHole<T>(T t)
+    {
+        // Store and format part as required
+        // Store name hole and value as KeyValuePair<string, object?>
+        return true;
+    }
+
+    public KeyValuePair<string, object?> this[int index]
+    {
+        get => ... // depends on the type used to store key value pairs
+    }
+
+    public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+    {
+        for (int i = 0; i < Count; ++i)
+        {
+            yield return this[i];
+        }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+}
+```
+
+### Q & A - The builder approach:
+
+- Question: Can I use a custom string interpolation builder in an async context?
+
+```csharp
+public async Task Method()
+{
+    // Cannot use MyStringBuilder if it's a ref struct
+    var key1 = "value1";
+    var key2 = "value2";
+    logger.LogInformation($"Some text: {key1}, {key2}");
```

Comment:
@stephentoub are there cases where this will falls back to a string.Format call? That would be bad.

Reply:
As long as the builder isn't a ref struct, you can expose a set of TryFormatXx methods so that it won't.